extensions/amp-story/0.1/amp-story-hint.js

```diff
@@ -0,0 +1,169 @@
+const NAVIGATION_HELP_OVERLAY = [
+  {
+    tag: 'div',
+    attrs: dict({'class': 'i-amphtml-story-navigation-help-overlay'}),
+    children: [
+      {
+        tag: 'div',
+        attrs: dict({'class': 'i-amphtml-story-navigation-help-section'
+            + ' prev-page'}),
+        children: [
+          {
+            tag: 'div',
+            attrs: dict({'class': 'i-amphtml-story-icons-container'}),
+            children: [
+              {
+                tag: 'div',
+                attrs: dict({'class': 'i-amphtml-story-hint-tap-icon'}),
+              },
+              {
+                tag: 'div',
+                attrs: dict({'class': 'i-amphtml-story-hint-tap-icon-text'}),
+              },
+            ],
+          },
+        ],
+      },
+      {
+        tag: 'div',
+        attrs: dict({'class': 'i-amphtml-story-navigation-help-section'
+            + ' next-page'}),
+        children: [
+          {
+            tag: 'div',
+            attrs: dict({'class': 'i-amphtml-story-icons-container'}),
+            children: [
+              {
+                tag: 'div',
+                attrs: dict({'class': 'i-amphtml-story-hint-tap-icon'}),
+              },
+              {
+                tag: 'div',
+                attrs: dict({'class': 'i-amphtml-story-hint-tap-icon-text'}),
+                text: 'Next page',
+              },
+            ],
+          },
+        ],
+      },
+    ],
+  },
+];
+
+/** @type {string} */
+const NAVIGATION_OVERLAY_CLASS = 'show-navigation-overlay';
+
+/** @type {string} */
+const FIRST_PAGE_OVERLAY_CLASS = 'show-first-page-overlay';
+
+/** @type {number} */
+const NAVIGATION_OVERLAY_TIMEOUT = 2000;
+
+/**
+ * User Hint Layer for <amp-story>.
+ */
+export class AmpStoryHint {
+
+  /**
+   * @param {!Window} win
+   */
+  constructor(win) {
+    /** @private {!Window} */
+    this.win_ = win;
+
+    /** @private {!Document} */
+    this.document_ = this.win_.document;
+
+    /** @private {?Element} */
+    this.hintContainer_ = null;
+
+    /** @private {?number} */
+    this.hintTimeout_ = null;
+  }
+
+  /**
+   * Builds the hint layer DOM.
+   * @return {!Element}
+   */
+  buildHintContainer() {
+    this.hintContainer_ = this.document_.createElement('aside');
+    this.hintContainer_.classList.add('i-amphtml-story-hint-container');
+    this.buildNavigationOverlay_();
+    return this.hintContainer_;
+  }
+
+  /**
+   * Builds navigation overlay DOM.
+   * @private
+   */
+  buildNavigationOverlay_() {
+    this.hintContainer_.appendChild(
+        renderSimpleTemplate(this.document_, NAVIGATION_HELP_OVERLAY));
+  }
+
+  /**
+   * Show navigation overlay DOM.
+   */
+  showNavigationOverlay() {
+    if (this.hintContainer_.classList.contains(NAVIGATION_OVERLAY_CLASS)) {
+      return;
+    }
+
+    this.hideAllNavigationHint();
+    this.hintContainer_.classList.add(NAVIGATION_OVERLAY_CLASS);
+
+    this.hintTimeout_ = setTimeout(() => {
```

Comment:
Should cancel any existing timeouts first, right?  Otherwise if this could happen:

- `t = 0ms` User triggers navigation overlay
- `t = 0ms` Navigation overlay is shown
- `t = 0ms` Schedule `this.hideAllNavigationHint()` to be executed at `t = 2000ms`
- `t = 1999ms` User triggers first page hint overlay
- `t = 1999ms` Navigation overlay is hidden
- `t = 1999ms` First page hint overlay is shown
- `t = 1999ms` Schedule `this.hideAllNavigationHint()` to be executed at `t = 3999ms`
- **`t = 2000ms` `this.hideAllNavigationHint()` is called (due to the timeout set from navigation overlay)**
- **`t = 2000ms` First page hint overlay is hidden**
- `t = 3999ms` `this.hideAllNavigationHint()` is called again (due to the timeout set from first page hint overlay)
- `t = 3999ms` (no-op)

*Result:*
The first page hint overlay is only shown for 1ms.

*Expected Result:*
The first page hint overlay should be shown for 2000ms.

Reply:
Debounce should the trick right?

Reply:
debounced it instead